server/models/image.py

```diff
@@ -114,3 +114,52 @@ def copyImage(self, image, recipe, creator=None):
             tags=['latest'], creator=creator, save=True, parent=image,
             description=image['description'], public=image['public'],
             config=image['config'], icon=image.get('icon', _DEFAULT_ICON))
+
+    def setAccessList(self, doc, access, save=False, user=None, force=False,
+                      setPublic=None, publicFlags=None):
+        """
+        Overrides AccessControlledModel.setAccessList to encapsulate ACL
+        functionality for an image.
+
+        :param doc: the image to set access settings on
+        :type doc: girder.models.image
+        :param access: The access control list
+        :type access: dict
+        :param save: Whether the changes should be saved to the database
+        :type save: bool
+        :param user: The current user
+        :param force: Set this to True to set the flags regardless of the passed in
+            user's permissions.
+        :type force: bool
+        :param setPublic: Pass this if you wish to set the public flag on the
+            resources being updated.
+        :type setPublic: bool or None
+        :param publicFlags: Pass this if you wish to set the public flag list on
+            resources being updated.
+        :type publicFlags: flag identifier str, or list/set/tuple of them,
+            or None
+        """
+        if setPublic is not None:
+            self.setPublic(doc, setPublic, save=False)
+
+        if publicFlags is not None:
+            doc = self.setPublicFlags(doc, publicFlags, user=user, save=False,
+                                      force=force)
+
+        doc = AccessControlledModel.setAccessList(self, doc, access,
+                                                  user=user, save=save, force=force)
+
+        recipeModels = AccessControlledModel.model('recipe', 'wholetale')
+
+        cursor = recipeModels.find({'_id': doc['recipeId']})
```

Comment:
instead of querying db directly we could use `load()` method here. Something along the lines:

```
recipe = self.model('recipe', 'wholetale').load(doc['recipeId'], user=user, level=AccessType.ADMIN)
```

It'll roughly do the same thing you've done below. We'll need to probably wrap it in `try: / except:` to catch invalid permission etc. and provide nice error message to the frontend.

Reply:
So I was under the assumption that if a user does not have permission to update the ACL of the recipe that the image was generated from, it should just silently continue without an error and only updates the image ACL. So are you suggesting that we should instead produce an error status if the user does not have permission to propagate the ACL from the image to the recipe? The only downside is if a user regularly generates images from other people's recipes they will see this error notification more often.

Reply:
Yeah, let's raise 403 for the time being. Reversing this to a silent failure will be easy from there. How exactly deal with the lack of permissions during a tale creation/publication is something that we should probably discuss during the All-hands meeting.

Reply:
👍 thanks for the clarification